druntime/src/core/internal/array/capacity.d

```diff
@@ -9,62 +9,150 @@
 */
 module core.internal.array.capacity;
 
-// HACK: `nothrow` and `pure` is faked.
-private extern (C) void[] _d_arraysetlengthT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
-private extern (C) void[] _d_arraysetlengthiT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
-
 /*
- * This template is needed because there need to be a `_d_arraysetlengthTTrace!Tarr` instance for every
- * `_d_arraysetlengthT!Tarr`. By wrapping both of these functions inside of this template we force the
- * compiler to create a instance of both function for every type that is used.
+ * Fully templated implementation of `_d_arraysetlengthT`, removing reliance on `TypeInfo`.
+ * The old `_d_arraysetlengthT` and `_d_arraysetlengthiT` functions have been removed.
+ */
+
+/**
+ * Resize a dynamic array by setting the `.length` property.
+ *
+ * Newly created elements are initialized to their default value.
+ *
+ * This function is **now fully templated**, eliminating the need for TypeInfo-based
+ * `_d_arraysetlengthT` and `_d_arraysetlengthiT`, while efficiently handling both
+ * zero-initialized and custom-initialized types.
+ *
+ * ---
+ * ## Example Usage:
+ * ```d
+ * void main()
+ * {
+ *     int[] a = [1, 2];
+ *     a.length = 3; // gets lowered to `_d_arraysetlengthT!(int[])(a, 3)`
+ * }
+ * ```
+ * ---
+ *
+ * - Uses a **templated approach** to minimize `TypeInfo` dependencies.
+ * - Follows `_d_newarrayU` for allocation and initialization.
+ * - Handles memory allocation and resizing safely, ensuring correct initialization.
  */
 
-/// Implementation of `_d_arraysetlengthT` and `_d_arraysetlengthTTrace`
-template _d_arraysetlengthTImpl(Tarr : T[], T)
+/// Complete templated implementation of `_d_arraysetlengthT` and its GC profiling variant `_d_arraysetlengthTTrace`
+
+size_t _d_arraysetlengthT(Tarr : T[], T)(
+    return scope ref Tarr arr,
+    size_t newlength,
+    string file = __FILE__,
+    int line = __LINE__,
+    string func = __FUNCTION__
+) @trusted
 {
-    private enum errorMessage = "Cannot resize arrays if compiling without support for runtime type information!";
+    import core.lifetime : emplace;
+    import core.internal.array.utils : __arrayAlloc;
+    import core.stdc.string : memcpy, memmove;
+    import core.internal.traits : Unqual;
 
-    /**
-     * Resize dynamic array
-     * Params:
-     *  arr = the array that will be resized, taken as a reference
-     *  newlength = new length of array
-     * Returns:
-     *  The new length of the array
-     * Bugs:
-     *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
-     */
-    size_t _d_arraysetlengthT(return scope ref Tarr arr, size_t newlength) @trusted pure nothrow
+    alias U = Unqual!T; // Ensure non-inout type
```

Comment:
Doing the `isMutable` trick from above will also allow you to use `T` instead of `U` as they'd be the same thing.